### PR TITLE
#163426535 pagination support for application

### PR DESCRIPTION
**Pull request by bobbylemm**
#### What does this PR do?
Implement pagination support for the application.

#### Description of the task to be completed?
Have the following endpoints working.
`post /api/v1/profiles?page=1`
`post /api/v1/articles/user/username?page=1`
`post /api/v1/articles?page=1`

#### How should this be manually tested?
After cloning the repo, run `npm install` and then `npm test`.

* Using Postman test the endpoints stated above.
  key: provide a `page number` and an optional `limit` in the request query

#### Relevant PT story ID?
[#163426535](https://www.pivotaltracker.com/story/show/163426535)

### CODE REVIEW

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by biolabalo — DISMISSED** [review]

**Review by omenkish — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Review by omenkish — DISMISSED** [review]

**Review by EfeAgare — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

File: server/controllers/articleController.js [hunk 1]
```diff
@@ -42,18 +44,18 @@ class ArticleController extends BaseHelper {
...
+          model: Users,
+          as: 'author',
+          attributes: ['username', 'bio', 'imageUrl']
+        },
+        { model: Comments },
+        { model: Reactions }
+      ];
+      const articles = await paginator(Articles, req, includedModels);
+      if (articles === undefined) return Response(res, 400, 'pagination error');
+      if (!articles.length) return Response(res, 200, 'No articles created yet');
```
Comment (p-funky) [inline server/controllers/articleController.js]: `No articles found` sounds more ideal and generic.

File: server/controllers/articleController.js [hunk 2]
```diff
@@ -72,13 +74,15 @@ class ArticleController extends BaseHelper {
...
       const { id: userId } = user;
-      const articles = await Articles.findAll({
-        where: { userId },
-        include: [{ model: Comments }, { model: Reactions }]
+      const includedModels = [{ model: Comments }, { model: Reactions }];
+      const articles = await paginator(Articles, req, includedModels, {
+        userId
       });
-      if (!articles.length) return errorResponse(res, 404, 'No articles found for user');
+      if (typeof articles === 'string') return Response(res, 200, articles);
```
Comment (p-funky) [inline server/controllers/articleController.js]: I thought we agreed this should no longer be present?

File: server/controllers/articleController.js [hunk 3]
```diff
@@ -42,18 +44,18 @@ class ArticleController extends BaseHelper {
...
+        {
+          model: Users,
+          as: 'author',
+          attributes: ['username', 'bio', 'imageUrl']
+        },
+        { model: Comments },
+        { model: Reactions }
+      ];
+      const articles = await paginator(Articles, req, includedModels);
+      if (articles === undefined) return Response(res, 400, 'pagination error');
```
Comment (p-funky) [inline server/controllers/articleController.js:57]: In what kind of cases can we get `undefined`?

File: server/controllers/userController.js [hunk 4]
```diff
@@ -202,10 +203,12 @@ class UserController extends BaseHelper {
...
   static async getAllUsersProfile(req, res, next) {
     try {
-      const users = await Users.findAll({
-        attributes: ['name', 'username', 'email', 'bio', 'imageUrl', 'createdAt']
-      });
-      return res.status(200).json(users);
+      const includedModels = [{ model: Articles }];
+      const profiles = await paginator(Users, req, includedModels);
+      if (!profiles) {
+        return res.status(200).json('There are no profiles for any user');
```
Comment (p-funky) [inline server/controllers/userController.js]: I think "There are no users to display" sounds more like it. But on the other hand, is this even possible?

File: server/helpers/paginator.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+) => {
+  const MAX = 8;
+  if (limit > MAX) {
+    limit = MAX;
+  }
+  const modifiedPage = page - 1;
+  const offset = parseInt(modifiedPage, 10) * limit;
+  if (page <= 0 || Number.isInteger(Number(modifiedPage) === false)) {
+    return undefined;
+  }
```
Comment (p-funky) [inline server/helpers/paginator.js:20]: Shouldn't the limit also be validated?

File: server/tests/app/followAndUnfollow.test.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  userBalogun,
+  userAkeem
+} from './mockData/userMockData';
+
+const { Users } = models;
+
+let tokenForBalogun;
+let tokenForBiola;
+// eslint-disable-next-line no-unused-vars
+let tokenForAkeem;
```
Comment (p-funky) [inline server/tests/app/followAndUnfollow.test.js]: Apparently, `tokenForAkeem` is not being used anywhere so we can as well delete it.

File: server/tests/app/followAndUnfollow.test.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+});
+
+
+before(async () => {
+  await Users.create(userAkeem);
+  return request(app)
+    .post('/api/v1/login')
+    .send({ email: userAkeem.email, password: userAkeem.password })
+    .then((res) => {
+      tokenForAkeem = res.body.token;
```
Comment (p-funky) [inline server/tests/app/followAndUnfollow.test.js]: Apparently, `tokenForAkeem` is not being used anywhere so we can as well delete it.

File: server/tests/app/followAndUnfollow.test.js (hunk 6 above)
Comment (biolabalo) [reply server/tests/app/followAndUnfollow.test.js]: It's actually being used but for some weird reason Eslint/HoundCI keeps flagging it

File: server/controllers/articleController.js (hunk 3 above)
Comment (bobbylemm) [reply server/controllers/articleController.js:57]: We get undefined when the validation on the input for the `page` and `limit` in the query fails.

File: server/controllers/userController.js (hunk 4 above)
Comment (bobbylemm) [reply server/controllers/userController.js]: Ok, I have implemented this feedback

File: server/helpers/paginator.js (hunk 5 above)
Comment (bobbylemm) [reply server/helpers/paginator.js:20]: I have implemented this feedback

File: server/tests/app/followAndUnfollow.test.js (hunk 7 above)
Comment (bobbylemm) [reply server/tests/app/followAndUnfollow.test.js]: It is actually being used

File: server/routes/userRoute.js [hunk 8]
```diff
@@ -20,7 +27,7 @@ const {
 router.post('/register', signUpValidator, registerUser);
 router.post('/login', loginUser);
 router.get('/profiles/:username', jwtAuth.authUser, getUserProfile);
-router.get('/profiles', jwtAuth.authUser, getAllUsersProfile);
+router.get('/profiles', getAllUsersProfile);
```
Comment (p-funky) [inline server/routes/userRoute.js]: Why is this route no longer authenticated?

File: server/tests/app/followAndUnfollow.test.js (hunk 7 above)
Comment (p-funky) [reply server/tests/app/followAndUnfollow.test.js]: Where?

File: server/routes/userRoute.js (hunk 8 above)
Comment (bobbylemm) [reply server/routes/userRoute.js]: Sorry for this, I probably took it out when I was refactoring the controller.
I have put it back though

**Review by biolabalo — APPROVED** [review]

**Review by divinediscipline — APPROVED** [review]
Looks good to me

**Review by omenkish — APPROVED** [review]

**Review by p-funky — APPROVED** [review]

**Pull request merged**